controllers/chaincode/chaincode_controller.go

```diff
@@ -222,77 +221,103 @@ func (r *ReconcileChaincode) ProposalUpdateFunc(e event.UpdateEvent) bool {
+			break
 		}
-		log.Info(fmt.Sprintf("proposal:%s done, chaincode status: %s", newProposal.GetName(), cr.Status.Phase))
+		if cond.Type == current.ProposalSucceeded {
+			cr.Status.Phase = current.ChaincodePhaseApproved
+			break
+		}
+	}
+	log.Info(fmt.Sprintf("proposal:%s done, chaincode status: %s", newProposal.GetName(), cr.Status.Phase))
+
+	if cr.Status.Phase != current.ChaincodePhaseApproved && cr.Status.Phase != current.ChaincodePhaseUnapproved {
+		log.Error(fmt.Errorf("expect %s or %s, but got %s",
+			current.ChaincodePhaseApproved, current.ChaincodePhaseUnapproved, cr.Status.Phase), "")
+		return false
+	}
+
+	if cr.Status.Phase == current.ChaincodePhaseApproved {
 		upgrade := false
-		if cr.Status.Phase == current.ChaincodePhaseApproved && newProposal.Spec.UpgradeChaincode != nil {
+		ccDef := newProposal.Spec.DeployChaincode
+		if ccDef == nil {
 			upgrade = true
-			if cr.Status.History == nil {
-				cr.Status.History = make([]current.ChaincodeHistory, 0)
-			}
-			shouldAppend := true
-			for _, item := range cr.Status.History {
+			ccDef = newProposal.Spec.UpgradeChaincode
+		}
+		if !upgrade && len(cr.Status.History) > 0 {
+			log.Error(fmt.Errorf("already deployed chaincode is not allowed to be redeployed"), "")
+			return false
+		}
+
+		image, digest, version, id, err := r.PickUpImageFromBuilder(ccDef.ExternalBuilder)
+		if err != nil {
+			log.Error(err, "the proposal passed, but failed to get the mirror information.")
+			return false
+		}
+		cr.Spec.ExternalBuilder = ccDef.ExternalBuilder
+		cr.Spec.Images.Name = image
+		cr.Spec.Images.Digest = digest
+		cr.Spec.Version = version
+		cr.Spec.ID = id
+
+		if err := r.client.Patch(context.TODO(), cr, nil, k8sclient.PatchOption{
+			Resilient: &k8sclient.ResilientPatch{
+				Retry:    3,
+				Into:     &current.Chaincode{},
+				Strategy: client.MergeFrom,
+			},
+		}); err != nil {
+			log.Error(err, fmt.Sprintf("failed to patch chaincode %s spec", cr.GetName()))
+			return false
+		}
+		// 需要重新获取
+		cr = &current.Chaincode{}
+		if err := r.client.Get(context.TODO(),
+			types.NamespacedName{Name: newProposal.Labels[current.ChaincodeProposalLabel]}, cr); err != nil {
+			log.Error(err, fmt.Sprintf("failed to get chaincode %s ", newProposal.Labels[current.ChaincodeProposalLabel]))
+			return false
+		}
+		cr.Status.Phase = current.ChaincodePhaseApproved
+		if upgrade {
+			appendHistory := true
+			for idx, item := range cr.Status.History {
 				if item.Image.Name == cr.Spec.Images.Name &&
 					item.Image.Digest == cr.Spec.Images.Digest &&
 					item.Image.PullSecret == cr.Spec.Images.PullSecret &&
 					item.Version == cr.Spec.Version {
-					shouldAppend = false
+					cr.Status.History[idx].UpgradeTime = v1.Now()
+					appendHistory = false
 					break
 				}
 			}
-			if shouldAppend {
+			if appendHistory {
 				cr.Status.History = append(cr.Status.History, current.ChaincodeHistory{
 					Version:     cr.Spec.Version,
```

Comment:
if external builder is used for deployment ,should cache `ExternalBuild` into history too.